File: core/src/main/java/javabot/operations/KarmaOperation.java
package javabot.operations;

import com.antwerkz.sofia.Sofia;
import javabot.Message;
import javabot.dao.KarmaDao;
import javabot.model.Karma;
import org.pircbotx.Channel;
import org.pircbotx.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.util.regex.Pattern;

public class KarmaOperation extends BotOperation {
    private static final Logger log = LoggerFactory.getLogger(KarmaOperation.class);

    private static final Pattern optionPattern = Pattern.compile("\\s--\\p{Alpha}[\\p{Alnum}]*=");

    @Inject
    private KarmaDao dao;

    @Override
    public boolean handleMessage(final Message event) {
        boolean handled = readKarma(event);
        if (!handled) {
            String message = event.getValue();
            final User sender = event.getUser();
            final Channel channel = event.getChannel();
            int operationPointer = message.indexOf("++");
            boolean increment = true;
            if (operationPointer == -1) {
                operationPointer = message.indexOf("--");
                increment = false;
                // check for no karma inc/dec, and ~-- and ~++ too
                if (operationPointer < 1) {
                    return false;
                }
            }

            /*
             * we won't get here unless operationPointer>0.
             *
             * But things get wonky; we need to handle two alternatives if it's a
             * karma decrement. One is: "admin --name=foo" and the other is
             * "foo --". We may need to apply a regex to ascertain whether
             * the signal is an option or not.
             *
             * The regex assumes options look like "--foo="
             */
            if (!increment) {
                final String potentialParam = message.substring(operationPointer - 1);
                if (optionPattern.matcher(potentialParam).find()) {
                    // we PRESUMABLY have an option...
                    return false;
                }
            }
            final String nick;
            try {
                nick = message.substring(0, operationPointer).trim().toLowerCase();
            } catch (StringIndexOutOfBoundsException e) {
                log.info("message = " + message, e);
                throw e;
            }
            // got an empty nick; spaces only?
            if (!nick.isEmpty()) {
                if (!channel.getName().startsWith("#")) {
                    getBot().postMessage(channel, event.getUser(), Sofia.privmsgChange(), event.isTell());
                    handled = true;
                } else {
                    if (nick.equalsIgnoreCase(sender.getNick())) {
                        if (increment) {
                            getBot().postMessage(channel, event.getUser(), Sofia.karmaOwnIncrement(), event.isTell());
                        }
                        increment = false;
                    }
                    Karma karma = dao.find(nick);
                    if (karma == null) {
                        karma = new Karma();
                        karma.setName(nick);
                    }
                    if (increment) {
                        karma.setValue(karma.getValue() + 1);
                    } else {
                        karma.setValue(karma.getValue() - 1);
                    }
                    karma.setUserName(sender.getNick());
                    dao.save(karma);
                    handled = readKarma(new Message(event.getChannel(), event.getUser(), "karma " + nick));
                }
            }
        }
        return handled;
    }

    public boolean readKarma(final Message event) {
           final String message = event.getValue();
           final Channel channel = event.getChannel();
           final User sender = event.getUser();
           if (message.startsWith("karma ")) {
               final String nick = message.substring("karma ".length()).toLowerCase();
               final Karma karma = dao.find(nick);
               if (karma != null) {
                   getBot().postMessage(channel, event.getUser(), nick.equalsIgnoreCase(sender.getNick())
                                                                  ? Sofia.karmaOwnValue(sender.getNick(), karma.getValue())
                                                                  : Sofia.karmaOthersValue(nick, karma.getValue(), sender.getNick())
                                           , event.isTell());
               } else {
                   getBot().postMessage(channel, event.getUser(), sender.getNick().equals(nick)
                                                                  ? Sofia.karmaOwnNone(sender.getNick())
                                                                  : Sofia.karmaOthersNone(nick, sender.getNick())
                                           , event.isTell());
               }
               return true;
   
           }
           return false;
       }

}

File: src/main/kotlin/javabot/operations/KarmaOperation.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import java.util.Locale
import java.util.regex.Pattern
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import javabot.dao.ChangeDao
import javabot.dao.ChannelDao
import javabot.dao.KarmaDao
import javabot.model.Karma
import javax.inject.Inject

class KarmaOperation
@Inject
constructor(
    bot: Javabot,
    adminDao: AdminDao,
    var dao: KarmaDao,
    var changeDao: ChangeDao,
    var channelDao: ChannelDao,
) : BotOperation(bot, adminDao) {

    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()

        readKarma(responses, event)
        if (responses.isEmpty()) {
            // if we're being addressed, we need to reconstruct the message so it looks like a
            // "karma operation"
            val message =
                if (event.addressed && !event.value.startsWith(bot.nick, true)) {
                    // argh. What if we're addressed, but it's not OUR KARMA being adjusted?
                    // We need to do some pre-emptive parsing to see if there's a match: if not,
                    // we have an "us++" situation.
                    val matches = operationPattern.matcher(event.value)
                    if (!matches.find()) {
                        "${bot.nick}${event.value}"
                    } else {
                        event.value
                    }
                } else {
                    event.value
                }
            val sender = event.user
            val channel = event.channel
            val matches = operationPattern.matcher(message)
            if (!matches.find()) {
                return responses
            }
            val tempSubject = matches.group(1)
            // we strip off the "completion character" if it's there
            val subject = tempSubject.substringBefore(":", tempSubject).trim()
            val operation = matches.group(2)
            var increment = operation.equals("++")

            // poison pill subject. Trust me on this one, people.
            if (subject.hashCode() == -1215158239) {
                return responses
            }

            if (event.tell) {
                responses.add(Message(event, Sofia.notAllowed()))
                return responses
            }

            // TODO handle tell (when there's an operation but no addressed nick: can we get here?)
            if (channel == null || !channel.name.startsWith("#")) {
                responses.add(Message(event, Sofia.privmsgChange()))
                return responses
            }
            if (increment && subject.equals(sender.nick, ignoreCase = true)) {
                responses.add(Message(event, Sofia.karmaOwnIncrement()))
                increment = false
            }
            val karma = dao.find(subject) ?: Karma(subject, 0, sender.nick)
            karma.value +=
                if (increment) {
                    1
                } else {
                    -1
                }
            dao.save(karma)
            changeDao.logKarmaChanged(
                karma.userName,
                karma.name,
                karma.value,
                channelDao.location(channel),
            )
            readKarma(responses, Message(event.channel, event.user, "karma $subject"))
        }
        return responses
    }

    private fun readKarma(responses: MutableList<Message>, event: Message) {
        val message = event.value
        val sender = event.user
        if (message.startsWith("karma ")) {
            val nick = message.substring("karma ".length).lowercase(Locale.getDefault())
            val karma = dao.find(nick)
            if (karma != null) {
                responses.add(
                    Message(
                        event,
                        if (nick.equals(sender.nick, ignoreCase = true))
                            Sofia.karmaOwnValue(sender.nick, karma.value)
                        else Sofia.karmaOthersValue(nick, karma.value, sender.nick),
                    )
                )
            } else {
                responses.add(
                    Message(
                        event,
                        if (sender.nick == nick) Sofia.karmaOwnNone(sender.nick)
                        else Sofia.karmaOthersNone(nick, sender.nick),
                    )
                )
            }
        }
    }

    companion object {
        private val operationPattern =
            Pattern.compile("^(?<nick>.+)(?<operation>\\+{2}|--).*\$", Pattern.COMMENTS)
    }
}
